### ISSUE TITLE
request for dataset access

### ISSUE THREAD

**Issue by VivianYuan12138** [houcharlie/PrE-Text#3]
Dear authors,

Thank you for sharing your excellent work on PrE-Text. I am very interested in reproducing your experiments and further exploring the results you presented in the paper. However, I noticed that the dataset used in your experiments (e.g., JOBS, FORUMS, MICROBLOG, and CODE) is not included in the repository. And also it is difficult to get them in c4_dataset.

Would it be possible for you to share the dataset or provide guidance on how to obtain or generate it?

**Comment by zhang-tuo-pdf** [comment]
Agree, could the author at least give an example for the data format?

**Comment by houcharlie** [comment]
Thank you both for your interest in our work! Sorry for the delay in response. 

Meta did not allow me to publish the details of the dataset during publication review (this was a paper done during an internship there). However, I can provide some code to help you try to reproduce the datasets:
```
from datasets import load_dataset
all_data = load_dataset('c4', 'en', split='train', streaming=True).shuffle(seed=100, buffer_size=1000000)
load_list = []
for i, text in enumerate(all_data):
    if [website url] in text['url']:
        load_list.extend(text['text'].split('\n'))
        if len(load_list) > 11000:
            break
train_data = load_list[:10000]
eval_data = {'1':load_list[10000:]}
```
Then save train_data and eval_data as jsons to a file location of your choice. For [website url] set the website url to something you choose. For initialization.json you can do a similar thing, except you filter out all the data in your "private datasets":
```
from datasets import load_dataset
all_data = load_dataset('c4', 'en', split='train', streaming=True).shuffle(seed=100, buffer_size=1000000)
load_list = []
for i, text in enumerate(all_data):
    if [website url] not in text['url']:
        load_list.extend(text['text'].split('\n'))
        if len(load_list) > 11000:
            break
initialization = load_list
```
Hope this is helpful! Let me know if there are any other questions.

**Comment by zhang-tuo-pdf** [comment]
Thank you so much! That is really helpful!

By the way, could you also share the shell file that you use to run the main.py?

**Comment by houcharlie** [comment]
It's in the readme:
```
TOKENIZERS_PARALLELISM=false accelerate launch main.py -outputdir $OUTPUT_DIR -cachedir \
$MODEL_DIR -datadir $DATASET_NAME -sensitivity $MAX_SAMPLES -sigma $NOISE -delta $DELTA
```

**Comment by zhang-tuo-pdf** [comment]
Thanks! I am also wondering where you define the number of clients. Are you using this line to do that?
https://github.com/houcharlie/PrE-Text/blob/main/nn_histogram.py#L40

**Comment by houcharlie** [comment]
Actually no, that code isn't really doing anything (leftover from when I was doing experiments on distributed machines). In PrE-Text you can simulate the federated setting by simply pooling the data from all the clients together into one training set (as long as you limit the number of samples per client). This is because having each client compute their histogram and then summing them up at the server is mathematically equivalent to pooling every client's data together and computing the histogram (as long as the number of samples per client is limited).

For our purposes this is sufficient because we are simulating the federated setting. If you were to implement PrE-Text in production you would need to implement logic for each client to independently construct a histogram and then sum them at the server.
